### DISCUSSION TITLE
Allow opting out of ZSH completions for Git

### DISCUSSION THREAD

**Discussion by seivan** [Homebrew/discussions#2797] [Tap maintenance and Homebrew development]
This [commit](https://github.com/Homebrew/homebrew-core/commit/f710a1395f44224e4bcc3518ee9c13a0dc850be1) removed the ability to opt out.

[This issue](https://github.com/Homebrew/homebrew-core/issues/13700) shows that this is inherently a degrading experience for the user. Some of the examples are posted [here](https://www.raphael-brugier.com/blog/fix-git-completion-zsh-mac-homebrew/) that documents some of the differences between the two approaches. 

There was [an attempt](https://github.com/Homebrew/homebrew-core/pull/52324) to completely remove the official ZSH completions that was shut down on the premise `So long as upstream are providing it: we'll be shipping it, sorry.`

However, this causes some [issues](https://github.com/Homebrew/homebrew-core/issues/33275)

Forcing the git completions on users, will degrade the experience. How much depends on what you expect, there has been some hyperbole, but in the end there is a degradation in user experience. 
There is no way out of it since the way `fpath` is setup, `site-functions` takes precedence. 
```
/usr/local/share/zsh/site-functions
/usr/local/Cellar/zsh/5.8_1/share/zsh/functions
```

1. git `<tab>` don't show aliases.
2. Don't properly work with other zsh plugins using git. 
3. Documentation on arguments are omitted.
And so on. 

Again, this is forced on a user, when installing `git` from Homebrew. 

One solution is for a manual removal of the symlink `/usr/local/share/zsh/site-functions/{_git, git-completion.bash}`, but that's not a good approach as `brew upgrade` will add them again. 
Another solution is to opt out of **all** Homebrew completions, this isn't a viable solution as most are actually useful and it's good to offer a nice out of the box experience. 

There was [PR](https://github.com/Homebrew/homebrew-core/pull/33157) that added the ability to opt out, but it was shut down in 2018. 
Can there be a new discussions to revive that, and allow users to opt out using `without-completion`

**Comment by carlocab** [comment]
Does prepending to `fpath` not work?

**Reply by seivan** [reply]
I am not sure what issues that will cause, so far I haven't touched or modified my fpath personally, maybe `OMZ` or `Zinit` is doing so. 
But I rather not modify my fpath and cause more issues down the line, if the alternative can be just opting out of git ZSH completion, which has been a point of discussion.

**Comment by h3xOo** [comment]
Homebrew *prepends* its zsh completion to the `fpath`, overwriting default zsh git completion. I hacked it by moving first entry of `fpath` at the end, right after `eval`ing Homebrew setup.

Here's the code I'm left with:
```zsh
eval "$(/opt/homebrew/bin/brew shellenv)"
# Move prepended Homebrew's fpath to the end.
fpath=(${fpath[@]:1} $fpath[1])
```

**Reply by CFD2** [reply]
what a legend, thank you!

It is a bit hard to believe nobody has done anything about it in so long

**Reply by rhcarvalho** [reply]
For future reference, I went with a more explicit take by removing the path by name:

```shell
if [[ -f "/opt/homebrew/bin/brew" ]]; then
  eval "$(/opt/homebrew/bin/brew shellenv)"
fi

# Remove the Homebrew completions that the above `brew shellenv` prepends to
# fpath, so that we can prefer system-level completions. In particular, the git
# completions that come with zsh are much better than the ones provided by the
# Homebrew formula, see https://github.com/orgs/Homebrew/discussions/2797.
fpath=("${fpath[@]:#/opt/homebrew/share/zsh/site-functions}")
# Now add the Homebrew site-functions back to the END of the list.
fpath+=(/opt/homebrew/share/zsh/site-functions(N-/))

```

**Comment by segevfiner** [comment]
Great, thanks to insistence on purism the Git formula actively harms me by degrading the completions quality, which has been annoying me for years, until I finally found out about this being the culprit. Not even documented in the Formula caveats, and the only opt out removed, cause I guess users couldn't even find it to use it in the first place so it wasn't used much...
